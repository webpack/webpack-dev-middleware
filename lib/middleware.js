'use strict';

const mime = require('mime');
const urlJoin = require('url-join');
const DevMiddlewareError = require('./DevMiddlewareError');
const { getFilenameFromUrl, handleRangeHeaders, handleRequest, ready } = require('./util');

module.exports = function wrapper(context) {
  return function middleware(req, res, next) {
    // fixes #282. credit @cexoso. in certain edge situations res.locals is
    // undefined.
    res.locals = res.locals || {};

    function goNext() {
      if (!context.options.serverSideRender) {
        return next();
      }

      return new Promise(((resolve) => {
        ready(context, () => {
          res.locals.webpackStats = context.webpackStats;
          resolve(next());
        }, req);
      }));
    }

    const acceptedMethods = context.options.methods || ['GET'];
    if (acceptedMethods.indexOf(req.method) === -1) {
      return goNext();
    }

    let filename = getFilenameFromUrl(context.options.publicPath, context.compiler, req.url);

    if (filename === false) {
      return goNext();
    }

    return new Promise(((resolve) => {
      handleRequest(context, filename, processRequest, req);
      function processRequest() {
        let notFound;
        try {
          let stat = context.fs.statSync(filename);

          if (!stat.isFile()) {
            if (stat.isDirectory()) {
              let { index } = context.options;

              if (index === undefined || index === true) {
                index = 'index.html';
              } else if (!index) {
                throw new DevMiddlewareError('next');
              }

              filename = urlJoin(filename, index);
              stat = context.fs.statSync(filename);
              if (!stat.isFile()) {
                throw new DevMiddlewareError('next');
              }
            } else {
              throw new DevMiddlewareError('next');
            }
          }
          notFound = false;
        } catch (e) {
          notFound = true;
        }
        const currentRequest = Object.assign({}, {
          url: req.url,
          method: req.method,
          headers: req.headers,
          filename,
          notFound
        });
        if (context.options && context.options.content && context.options.content.filter) {
          if (!context.options.content.filter(currentRequest)) {
            return goNext();
          }
        }
        const defaultTransform = (localRequest, fs) => new Promise((localResolve) => {
          if (notFound) {
            return localResolve(null);
          }
          fs.readFile(localRequest.filename, (err, content) => {
            if (err) {
              localResolve(null);
            } else {
              localResolve(content);
            }
          });
        });
        const optionTransform = (context.options && context.options.content) ? context.options.content.transform : null;
        return defaultTransform(currentRequest, context.fs)
          .then(content => ((!optionTransform) ? content : optionTransform(currentRequest, content, context.fs)))
          .then((content) => {
            if (content == null) {
              return Promise.reject();
            }
            content = handleRangeHeaders(content, req, res);
            let contentType = mime.getType(currentRequest.filename);

            // do not add charset to WebAssembly files, otherwise compileStreaming will fail in the client
            if (!/\.wasm$/.test(currentRequest.filename)) {
              contentType += '; charset=UTF-8';
            }

            res.setHeader('Content-Type', contentType);
            res.setHeader('Content-Length', content.length);

            const { headers } = context.options;
            if (headers) {
              for (const name in headers) {
                if ({}.hasOwnProperty.call(headers, name)) {
                  res.setHeader(name, context.options.headers[name]);
                }
              }
            }
            // Express automatically sets the statusCode to 200, but not all servers do (Koa).
            res.statusCode = res.statusCode || 200;
            if (res.send) res.send(content);
            else res.end(content);
            resolve();
          })
          .catch(goNext);
      }
    }));
  };
};
