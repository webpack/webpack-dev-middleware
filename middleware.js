/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
var path = require("path");
var MemoryFileSystem = require("memory-fs");
var mime = require("mime");

// constructor for the middleware
module.exports = function(compiler, options) {
	if(!options) options = {};
	if(options.watchDelay === undefined) options.watchDelay = 200;
	if(typeof options.stats === "undefined") options.stats = {};
	if(!options.stats.context) options.stats.context = process.cwd();

	// store our files in memory
	var files = {};
	var fs = compiler.outputFileSystem = (options.useRealFileSystem && compiler.outputFileSystem) || new MemoryFileSystem();

	compiler.plugin("done", function(stats) {
		// We are now on valid state
		state = true;
		// Do the stuff in nextTick, because bundle may be invalidated
		//  if a change happend while compiling
		process.nextTick(function() {
			// check if still in valid state
			if(!state) return;
			// print webpack output
			var displayStats = (!options.quiet && options.stats !== false);
			if(displayStats &&
				!(stats.hasErrors() || stats.hasWarnings()) &&
				options.noInfo)
				displayStats = false;
			if(displayStats) {
				console.log(stats.toString(options.stats));
			}
			if (!options.noInfo && !options.quiet)
				console.info("webpack: bundle is now VALID.");

			// execute callback that are delayed
			var cbs = callbacks;
			callbacks = [];
			cbs.forEach(function continueBecauseBundleAvailible(cb) {
				cb();
			});
		});

		// In lazy mode, we may issue another rebuild
		if(forceRebuild) {
			forceRebuild = false;
			rebuild();
		}
	});

	// on compiling
	function invalidPlugin() {
		if(state && (!options.noInfo && !options.quiet))
			console.info("webpack: bundle is now INVALID.");
		// We are now in invalid state
		state = false;
	}
	compiler.plugin("invalid", invalidPlugin);
	compiler.plugin("compile", invalidPlugin);

	// the state, false: bundle invalid, true: bundle valid
	var state = false;

	// in lazy mode, rebuild automatically
	var forceRebuild = false;

	// delayed callback
	var callbacks = [];

	// wait for bundle valid
	function ready(fn, req) {
		if(state) return fn();
		if(!options.noInfo && !options.quiet)
			console.log("webpack: wait until bundle finished: " + req.url);
		callbacks.push(fn);
	}

	// start watching
	if(!options.lazy) {
		var watching = compiler.watch(options.watchDelay, function(err) {
			if(err) throw err;
		});
	} else {
		state = true;
	}

	function rebuild() {
		if(state) {
			state = false;
			compiler.run(function(err) {
				if(err) throw err;
			});
		} else {
			forceRebuild = true;
		}
	}

	function pathJoin(a, b) {
		return a == "/" ? "/" + b : (a||"") + "/" + b
	}

	function getFilenameFromUrl(url) {
		// publicPrefix is the folder our bundle should be in
		var localPrefix = options.publicPath || "/";
		if(url.indexOf(localPrefix) !== 0) {
			if(/^(https?:)?\/\//.test(localPrefix)) {
				localPrefix = "/" + localPrefix.replace(/^(https?:)?\/\/[^\/]+\//, "");
				// fast exit if another directory requested
				if(url.indexOf(localPrefix) !== 0) return false;
			} else return false;
		}
		// get filename from request
		var filename = url.substr(localPrefix.length);
		if(filename.indexOf("?") >= 0) {
			filename = filename.substr(0, filename.indexOf("?"));
		}
		return filename ? pathJoin(compiler.outputPath, filename) : compiler.outputPath;
	}

	// The middleware function
	function webpackDevMiddleware(req, res, next) {
		var filename = getFilenameFromUrl(req.url);
		if (filename === false) return next();

		// in lazy mode, rebuild on bundle request
		if(options.lazy && filename === pathJoin(compiler.outputPath, options.filename))
			rebuild();
		// delay the request until we have a vaild bundle
		ready(function() {
			try {
				var stat = fs.statSync(filename);
				if(!stat.isFile()) {
					if (stat.isDirectory()) {
						filename = path.join(filename, "index.html");
						stat = fs.statSync(filename);
						if(!stat.isFile()) throw "next";
					} else {
						throw "next";
					}
				}
			} catch(e) {
				return next();
			}

			// server content
			var content = fs.readFileSync(filename);
			res.setHeader("Access-Control-Allow-Origin", "*"); // To support XHR, etc.
			res.setHeader("Content-Type", mime.lookup(filename));
			res.setHeader("Content-Length", content.length);
			if(options.headers) {
				for(var name in options.headers) {
					res.setHeader(name, options.headers[name]);
				}
			}
			res.end(content);
		}, req);
	}

	webpackDevMiddleware.getFilenameFromUrl = getFilenameFromUrl;

	webpackDevMiddleware.invalidate = function() {
		if(watching) watching.invalidate();
	};
	webpackDevMiddleware.close = function(callback) {
		callback = callback || function(){};
		if(watching) watching.close(callback);
		else callback();
	};

	webpackDevMiddleware.fileSystem = fs;

	return webpackDevMiddleware;
}
