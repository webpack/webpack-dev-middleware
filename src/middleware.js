import path from 'path';

import mime from 'mime';

import DevMiddlewareError from './DevMiddlewareError';
import getFilenameFromUrl from './utils/getFilenameFromUrl';
import handleRangeHeaders from './utils/handleRangeHeaders';
import ready from './utils/ready';

// Do not add a charset to the Content-Type header of these file types
// otherwise the client will fail to render them correctly.
const nonCharsetFileTypes = /\.(wasm|usdz)$/;

const HASH_REGEXP = /[0-9a-f]{10,}/;

export default function wrapper(context) {
  return function middleware(req, res, next) {
    // fixes #282. credit @cexoso. in certain edge situations res.locals is
    // undefined.
    // eslint-disable-next-line no-param-reassign
    res.locals = res.locals || {};

    function goNext() {
      if (!context.options.serverSideRender) {
        return next();
      }

      return new Promise((resolve) => {
        ready(
          context,
          () => {
            // eslint-disable-next-line no-param-reassign
            res.locals.webpack = {
              stats: context.stats,
              outputFileSystem: context.outputFileSystem,
            };

            resolve(next());
          },
          req
        );
      });
    }

    const acceptedMethods = context.options.methods || ['GET', 'HEAD'];

    if (acceptedMethods.indexOf(req.method) === -1) {
      return goNext();
    }

    let filename = getFilenameFromUrl(context, req.url);

    if (filename === false) {
      return goNext();
    }

    return new Promise((resolve) => {
      // eslint-disable-next-line consistent-return
      function processRequest() {
        try {
          let stat = context.outputFileSystem.statSync(filename);

          if (!stat.isFile()) {
            if (stat.isDirectory()) {
              let { index } = context.options;

              // eslint-disable-next-line no-undefined
              if (index === undefined || index === true) {
                index = 'index.html';
              } else if (!index) {
                throw new DevMiddlewareError('next');
              }

              filename = path.posix.join(filename, index);
              stat = context.outputFileSystem.statSync(filename);

              if (!stat.isFile()) {
                throw new DevMiddlewareError('next');
              }
            } else {
              throw new DevMiddlewareError('next');
            }
          }
        } catch (_ignoreError) {
          return resolve(goNext());
        }

        // server content
        let content;

        try {
          content = context.outputFileSystem.readFileSync(filename);
        } catch (_ignoreError) {
          return resolve(goNext());
        }

        content = handleRangeHeaders(content, req, res);

        let contentType = mime.getType(filename) || '';

        if (!nonCharsetFileTypes.test(filename)) {
          contentType += '; charset=UTF-8';
        }

        if (!res.get('Content-Type')) {
          res.set('Content-Type', contentType);
        }

        const { headers } = context.options;

        if (headers) {
          for (const name in headers) {
            if ({}.hasOwnProperty.call(headers, name)) {
              res.setHeader(name, context.options.headers[name]);
            }
          }
        }

        res.send(content);

        resolve();
      }

      if (HASH_REGEXP.test(filename)) {
        try {
          if (context.outputFileSystem.statSync(filename).isFile()) {
            processRequest();

            return;
          }
        } catch (_ignoreError) {
          // Ignore error
        }
      }

      ready(context, processRequest, req);
    });
  };
}
