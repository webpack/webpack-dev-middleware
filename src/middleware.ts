import path from 'path';

import type express from 'express';
import mime from 'mime-types';

import getFilenameFromUrl from './utils/getFilenameFromUrl';
import handleRangeHeaders from './utils/handleRangeHeaders';
import ready from './utils/ready';
import type { WebpackDevMiddlewareContext } from './types';

export default function wrapper(
  context: WebpackDevMiddlewareContext
): express.RequestHandler {
  return async function middleware(req, res, next) {
    const acceptedMethods = context.options.methods || ['GET', 'HEAD'];
    // fixes #282. credit @cexoso. in certain edge situations res.locals is undefined.
    res.locals = res.locals || {};

    if (!acceptedMethods.includes(req.method)) {
      await goNext();
      return;
    }

    // eslint-disable-next-line @typescript-eslint/no-misused-promises
    ready(context, processRequest, req);

    async function goNext() {
      if (!context.options.serverSideRender) {
        return next();
      }

      return new Promise((resolve) => {
        ready(
          context,
          () => {
            res.locals.webpack = { devMiddleware: context };
            resolve(next());
          },
          req
        );
      });
    }

    async function processRequest() {
      const filename = getFilenameFromUrl(context, req.url);
      const { headers } = context.options;
      let content: Buffer;

      if (!filename) {
        await goNext();
        return;
      }

      try {
        content = context.outputFileSystem.readFileSync(filename);
      } catch (_ignoreError) {
        await goNext();
        return;
      }

      if (!res.get('Content-Type')) {
        // content-type name(like application/javascript; charset=utf-8) or false
        const contentType = mime.contentType(path.extname(filename));

        if (contentType) {
          res.set('Content-Type', contentType);
        }
      }

      if (headers) {
        for (const name of Object.keys(headers)) {
          res.set(name, headers[name]);
        }
      }

      content = handleRangeHeaders(context, content, req, res);
      res.send(content);
    }
  };
}
