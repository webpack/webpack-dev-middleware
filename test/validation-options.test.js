import path from "path";

import { createFsFromVolume, Volume } from "memfs";

import middleware from "../src";

import getCompiler from "./helpers/getCompiler";

// Suppress unnecessary stats output
global.console.log = jest.fn();

const configuredFs = createFsFromVolume(new Volume());

configuredFs.join = path.join.bind(path);

describe("validation", () => {
  const cases = {
    mimeTypes: {
      success: [{ phtml: ["text/html"] }],
      failure: ["foo"],
    },
    writeToDisk: {
      success: [true, false, () => {}],
      failure: [{}],
    },
    methods: {
      success: [["GET", "HEAD"]],
      failure: [{}, true],
    },
    headers: {
      success: [
        { "X-Custom-Header": "yes" },
        () => {},
        [{ key: "foo", value: "bar" }],
      ],
      failure: [true, 1, [], [{ foo: "bar" }]],
    },
    publicPath: {
      success: ["/foo", "", "auto", () => "/public/path"],
      failure: [false],
    },
    serverSideRender: {
      success: [true],
      failure: ["foo", 0],
    },
    outputFileSystem: {
      success: [configuredFs],
      failure: [false],
    },
    index: {
      success: [true, false, "foo"],
      failure: [0, {}],
    },
    stats: {
      success: [true, false, "normal", "verbose", { all: false, assets: true }],
      failure: [0],
    },
    mimeTypeDefault: {
      success: ["text/plain"],
      failure: [0],
    },
    modifyResponseData: {
      success: [
        (_ignore, _ignore1, foo, bar) => {
          return { foo, bar };
        },
      ],
      failure: [true],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === "object" && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  async function createTestCase(key, value, type) {
    it(`should ${
      type === "success" ? "successfully validate" : "throw an error on"
    } the "${key}" option with "${stringifyValue(value)}" value`, (done) => {
      const compiler = getCompiler();

      let webpackDevMiddleware;
      let error;

      try {
        webpackDevMiddleware = middleware(compiler, { [key]: value });
      } catch (maybeError) {
        if (maybeError.name !== "ValidationError") {
          throw maybeError;
        }

        error = maybeError;
      } finally {
        if (type === "success") {
          expect(error).toBeUndefined();
        } else if (type === "failure") {
          expect(() => {
            throw error;
          }).toThrowErrorMatchingSnapshot();
        }

        if (webpackDevMiddleware) {
          webpackDevMiddleware.waitUntilValid(() => {
            webpackDevMiddleware.close(() => {
              done();
            });
          });
        } else {
          done();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(cases)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
