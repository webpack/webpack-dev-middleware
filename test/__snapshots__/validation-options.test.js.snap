// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation should throw an error on the "headers" option with "true" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.headers should be an object:
   object { … }"
`;

exports[`validation should throw an error on the "index" option with "{}" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.index should be one of these:
   boolean | string
   Details:
    * configuration.index should be a boolean.
    * configuration.index should be a string."
`;

exports[`validation should throw an error on the "index" option with "0" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.index should be one of these:
   boolean | string
   Details:
    * configuration.index should be a boolean.
    * configuration.index should be a string."
`;

exports[`validation should throw an error on the "methods" option with "{}" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.methods should be an array:
   [string, ...]"
`;

exports[`validation should throw an error on the "methods" option with "true" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.methods should be an array:
   [string, ...]"
`;

exports[`validation should throw an error on the "mimeTypes" option with "foo" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.mimeTypes should be an object:
   object { … }"
`;

exports[`validation should throw an error on the "outputFileSystem" option with "false" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.outputFileSystem should be an object:
   object { … }"
`;

exports[`validation should throw an error on the "publicPath" option with "false" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.publicPath should be a string."
`;

exports[`validation should throw an error on the "serverSideRender" option with "0" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.serverSideRender should be a boolean."
`;

exports[`validation should throw an error on the "serverSideRender" option with "foo" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.serverSideRender should be a boolean."
`;

exports[`validation should throw an error on the "stats" option with "0" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.stats should be one of these:
   object { … } | boolean | \\"none\\" | \\"errors-only\\" | \\"errors-warnings\\" | \\"minimal\\" | \\"normal\\" | \\"verbose\\"
   Details:
    * configuration.stats should be an object:
      object { … }
    * configuration.stats should be a boolean.
    * configuration.stats should be one of these:
      \\"none\\" | \\"errors-only\\" | \\"errors-warnings\\" | \\"minimal\\" | \\"normal\\" | \\"verbose\\""
`;

exports[`validation should throw an error on the "stats" option with "foo" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.stats should be one of these:
   object { … } | boolean | \\"none\\" | \\"errors-only\\" | \\"errors-warnings\\" | \\"minimal\\" | \\"normal\\" | \\"verbose\\"
   Details:
    * configuration.stats should be an object:
      object { … }
    * configuration.stats should be a boolean.
    * configuration.stats should be one of these:
      \\"none\\" | \\"errors-only\\" | \\"errors-warnings\\" | \\"minimal\\" | \\"normal\\" | \\"verbose\\""
`;

exports[`validation should throw an error on the "watchOptions" option with "0" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.watchOptions should be an object:
   object { … }"
`;

exports[`validation should throw an error on the "writeToDisk" option with "{}" value 1`] = `
"Invalid configuration object. Object has been initialized using a configuration object that does not match the API schema.
 - configuration.writeToDisk should be one of these:
   boolean | function
   Details:
    * configuration.writeToDisk should be a boolean.
    * configuration.writeToDisk should be an instance of function."
`;
